pykafka/producer.py

```diff
@@ -111,15 +158,112 @@ def __repr__(self):
             id_=hex(id(self))
         )
 
-    def _send_request(self, broker, req, attempt):
+    def __enter__(self):
+        """Context manager entry point - start the producer"""
+        self.start()
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Context manager exit point - stop the producer
+
+        If __exit__ is being called as a result of self._worker_exception
+        being raised, don't wait on the workers to finish. If __exit__ is being
+        called because the with: block has ended, do wait for workers to finish.
+        """
+        # If the thread crashed, don't wait for it
+        self.stop(wait=self._worker_exception is None)
+
+    @raise_worker_exceptions
+    def start(self):
+        """Set up data structures and start worker threads"""
+        if not self._running:
+            self._owned_brokers = {}
+            for partition in self._topic.partitions.values():
+                if partition.leader.id not in self._owned_brokers:
+                    self._owned_brokers[partition.leader.id] = OwnedBroker(
+                        self, partition.leader)
+            self._running = True
+
+    @raise_worker_exceptions
```

Comment:
I'm not sure this should raise exceptions, actually. If it raises because one worker is dead, how will the others be notified they should stop?